provisioner/aws.go

```diff
@@ -203,24 +206,65 @@ func (a *awsAdapter) CreateOrUpdateClusterStack(stackName, stackDefinitionPath s
 		return nil, fmt.Errorf("'%s' config item is missing, must be defined", workerSharedSecretConfigItemKey)
 	}
 
-	workerNodes := workerPool.MinSize
+	workerPoolDesired := workerPool.MinSize
+	workerPoolMinSize := workerPool.MinSize
+	workerPoolMaxSize := workerPool.MaxSize
+	workerPoolInstanceType := workerPool.InstanceType
+
+	// we currently don't support scaling for master pools
+	if masterPool.MinSize != masterPool.MaxSize {
+		return nil, fmt.Errorf("master pool must have the same min_size and max_size")
+	}
+
+	masterPoolDesired := masterPool.MinSize
+	masterPoolInstanceType := masterPool.InstanceType
 
 	// if stack already exists use the already generated secret and current
 	// desired worker nodes.
 	if stack != nil {
+		desiredCapacity := int64(0)
+
 		// get desired worker nodes
-		asg, err := a.getNodePoolASG(stackName, workerPool.Name)
+		asg, err := a.getNodePoolASG(cluster.ID, workerPool.Name)
 		if err != nil {
-			return nil, err
+			if err != ErrASGNotFound {
+				return nil, err
+			}
+		} else {
+			desiredCapacity = aws.Int64Value(asg.DesiredCapacity)
 		}
 
 		// reduce the desired size if necessary
-		workerNodes = int64(math.Min(float64(*asg.DesiredCapacity), float64(workerPool.MaxSize)))
+		workerPoolDesired = int64(math.Min(float64(desiredCapacity), float64(workerPool.MaxSize)))
+
+		// TODO(tech-depth): this is only used for migrating from
+		// legacy node pools to real node pool support.
+		if nodePoolFeatureEnabled(cluster) {
+			workerPoolDesired = aws.Int64Value(asg.DesiredCapacity)
+			workerPoolMinSize = aws.Int64Value(asg.MinSize)
+			workerPoolMaxSize = workerPoolDesired
+
+			// get master nodepool size
+			asg, err := a.getNodePoolASG(cluster.ID, masterPool.Name)
+			if err != nil {
+				if err != ErrASGNotFound {
+					return nil, err
+				}
```

Comment:
if err = ErrASGNotFound we create a worker ASG with 0 desired, does it makes sense?

Reply:
This is intended for the legacy node pool migration. I added a comment to make it more clear.